Add OuputVerbosity and use it in assertions

plannigan opened the pull request:
<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [X] Include documentation when adding new features.
- [X] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [ ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [X] Add yourself to `AUTHORS` in alphabetical order.
-->

`OuputVerbosity` is a new class that allows for retrieving a verbosity level for a specific output type. If nothing is specifically configured, the application wide verbose value is used.

The first use of this functionality is `verbosity_assertions` which can be used to make the assertion reports output at the most verbose level while still showing a single character for each test case when progress is being reported.


Addresses #11387

plannigan commented on src/_pytest/config/__init__.py:
```diff
@@ -1662,6 +1663,54 @@ def _warn_about_skipped_plugins(self) -> None:
+        return f"verbosity_{output_type}"
+
+    @staticmethod
+    def add_ini(parser: "Parser", output_type: str, help: str) -> None:
```
Since many things will add these type of settings, I thought it would be helpful to have this function to ensure it is dong consistently. I'm not completely in love with this level of indirection. But it is necessary since config settings are added to the `Parser` before the `Config` object exists.

nicoddemus replied:
Good idea. 👍

plannigan commented on testing/test_assertion.py:
```diff
@@ -17,12 +17,23 @@
-    class Config:
-        def getoption(self, name):
-            if name == "verbose":
+def mock_config(verbose: int = 0, assertion_override: Optional[int] = None):
```
With the move away from `getoption()`, this needed to get a bit more complicated. (In the long term, it would probably be beneficial move the code base to a consistent way to access the verbosity level.)

plannigan commented:
Since this is my first contribution, I'm not sure the best way to get a review for this PR that has been up for 3 weeks. @nicoddemus & @Zac-HD, you both interacted with the issue. Would one of you be able to review this code?

nicoddemus commented:
Sorry @plannigan, you are not supposed to do anything else other than opening the PR: we dropped the ball for not reviewing it earlier (I personally have been busy). When that happens a ping (like you did) is appreciated, thank you!

I will try to review this in the next couple of days.

nicoddemus commented on doc/en/how-to/output.rst:
```diff
@@ -270,6 +270,16 @@ situations, for example you are shown even fixtures that start with ``_`` if you
+In addition to specifying the application wide verbosity level, it is possible to control specific aspects independently.
+This is done by setting a verbosity level in the configuration file for the specific aspect of the output.
+
+``verbosity_assertions``: Controls how verbose the assertion output should be when pytest is executed. A value of ``2``
```
```suggestion
:confval:`verbosity_assertions`: Controls how verbose the assertion output should be when pytest is executed. A value of ``2``
```

nicoddemus commented on doc/en/how-to/output.rst:
```diff
@@ -270,6 +270,16 @@ situations, for example you are shown even fixtures that start with ``_`` if you
+``verbosity_assertions``: Controls how verbose the assertion output should be when pytest is executed. A value of ``2``
+would have the same output as the previous example, but each test inside the file is shown by a single character in the
+output.
+
```
```suggestion

(Note: currently this is the only option available, but more might be added in the future).
```

nicoddemus commented on src/_pytest/assertion/__init__.py:
```diff
@@ -42,6 +43,14 @@ def pytest_addoption(parser: Parser) -> None:
+        "assertions",
+        help=(
+            "Specify a verbosity level for assertions, overriding the main level. "
+            "Higher levels will provide more a more detailed explanation when an assertion fails."
```
```suggestion
            "Higher levels will provide more detailed explanation when an assertion fails."
```

nicoddemus commented on src/_pytest/config/__init__.py:
```diff
@@ -1662,6 +1663,54 @@ def _warn_about_skipped_plugins(self) -> None:
+        """Application wide verbosity level."""
+        return cast(int, self._config.option.verbose)
+
+    def verbosity_for(self, output_type: str) -> int:
```
I think if one passes an invalid `output_type` at the moment, we will get an error (dur to `self._config.getini`).

We probably want to have this classs accessible to plugins and users, via `config.output_verbosity`, so we need to think about its API carefully as we will need to support this officially moving on.

Some points:

1. In tandem with pytest's direction regarding type safety, I think we should use an `Enum` with the output types currently available, something along the lines of:

   ```python
   class VerbosityType(Enum):
       """..."""     

       Global = "global"
       Assertions = "assertions"
   ```

   Instead of a simple `output_type: str` parameter/value.

2. The class name is `OutputVerbosity` and the public methods are `verbosity` and `verbosity_for`, which read a bit redundant: `confg.output_verbosity.verbosity`, `confg.output_verbosity.get_verbosity_for`. I suggest we use simply `get`, defaulting to the `Global` verbosity:

   ```python
   def get(self, type: VerbosityType = VerbosityType.Global) -> int:
       ...
   ```

   Which reads well I think:

   ```python
   config.output_verbosity.get()
   config.output_verbosity.get(VerbosityType.Assertions)
   ```

3. We need to properly document both `OutputVerbosity` and `VerbosityType` in the reference docs (including some usage examples, but possibly we can write those in the class docstring). We also need to expose `VerbosityType` in the `pytest` namespace.

4. Given `OutputVerbosity` is public, we should make `OutputVerbosity.addini` private, as it is intended for internal use only.


I'm not married to the names above, so suggestions and bikeshedding are welcome.

plannigan replied:
1. I opted for `str` because I was thinking that plugins might want also register types, like any other config setting. Which would mean that it wouldn't be possible to statically define all of the types. However, reflecting on that, I'm not sure that level of extensibility is necessary. It is likely better for `pytest` to say "here are N buckets that a plugin can utilize" and that way users of multiple plugins that leverage this functionality will see consistent behavior without needing to enable M settings.
2. That is fair. It also works nicely with the enum suggestion.
3. Sure
4. I had it as public for the reasons described in 1, but it makes sense to make it private based on that change.

nicoddemus commented on changelog/11387.feature.rst:
```diff
@@ -0,0 +1 @@
+:confval:`verbosity_assertions` option added to be able to control assertion output independent of the application wide verbosity level.
```
We should also mention `OutputVerbosity` and link to its docs.

nicoddemus reviewed (changes requested):
Great work @plannigan, and sorry again for the delay!

The implementation is good, however I think we might need some more interactions given this introduce a new API, which must be done carefully and with care.

bluetech reviewed (commented):
I'm not able to do a full review ATM, but left some quick comments.

plannigan commented on doc/en/how-to/output.rst:
```diff
@@ -270,6 +270,16 @@ situations, for example you are shown even fixtures that start with ``_`` if you
+Fine gain verbosity
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In addition to specifying the application wide verbosity level, it is possible to control specific aspects independently.
```
My intention is to implement some of the other verbosity types I identified once this PR is merged. A different suggestion was to add a note that only one is currently supported. That seems sufficient to me if "life happens" and it takes me some time to get to adding other types.

plannigan commented on doc/en/reference/reference.rst:
```diff
@@ -1819,6 +1819,19 @@ passed multiple times. The expected format is ``name=value``. For example::
             clean_db
 
 
+.. confval:: verbosity_assertions
```
Two reasons:

* My use case is to always enable `-vv` for assertions without increasing the verbosity of the rest of the output. Setting that in the configuration file allows this to be enabled for a project without being explicitly added each time `pytest` is run.
* My initial implementation took an arbitrary string when adding a new level. The thought was that any plugin might add their own type. This would mean that there could potentially be a large number of command line options related to fine-grained verbosity, which seemed harder to manage. With the move to `pytest` statically defining each type, that will keep the N small, so I'm less apposed to the idea.
* With the existence of `addopts` and `--override-ini`, it isn't clear if there are rules about what should be a command line option version a configuration file setting.

nicoddemus replied:
I agree this is something one might want to configure mainly via the configuration file, so a configuration value seems reasonable.

plannigan commented on src/_pytest/config/__init__.py:
```diff
@@ -1662,6 +1663,54 @@ def _warn_about_skipped_plugins(self) -> None:
             )
 
 
+class OutputVerbosity:
```
I'm open tweaking the name. I generally default to more explicit names.

plannigan commented on src/_pytest/config/__init__.py:
```diff
@@ -1662,6 +1663,54 @@ def _warn_about_skipped_plugins(self) -> None:
+        self._config = config
+
+    @property
+    def verbose(self) -> int:
```
Like the class name, I'm open tweaking the name. I went with `verbose` to make the command line option name destination (which is how most things access the application wide value now). But `pytest` is already inconsistent:

* command line options name: `--verbosity`
* command line options destination name: `verbose`
* `TerminalReporter` property: `verbosity`

Maybe a follow up cleanup PR would be to make the code base more consistent.

nicoddemus replied:
I suggest we just remove this option, and let users access it via `.get()`.

Note: I removed and pushed this property in a separate commit, we can revert it if we decide it otherwise.

plannigan commented:
I believe I have addressed the technical comments and documentation requests.

I'm not sure why the doc building step is failing. The error don't seem to be related to the lines I changed. If someone with more experience with `sphinx` can take a look, I would appreciate it.

nicoddemus commented:
> I'm not sure why the doc building step is failing

The problem was that you used `^^^` for the heading in your new session, while in the rest of the document we used `~~~` for that level of heading. I pushed a fix. 👍

bluetech reviewed (changes requested):
Thanks for the update @plannigan, my comments:

1. The `OutputVerbosity` and `VerbosityType` seem more complicated than necessary to me. What is the advantage over doing it the direct way, i.e. adding the ini directly and maybe adding a getter on `Config`?

2. If we stick with `OutputVerbosity`, let's make it just `Verbosity` -- the `Output` part is entirely redundant IMO.

3. We should have at least one end-to-end test (`pytester.runpytest` test) which shows this working, i.e. changing the output vs. not setting the ini.

nicoddemus commented:
> i.e. adding the ini directly 

I think this is to handle the fallback to default verbosity when the setting is defined in the ini file.

> and maybe adding a getter on Config?

Can you clarify what you mean here, do you mean adding:

1. `Config.get_verbosity(VerbosityType) -> int` 
2. `Config.get_verbosity(Literal["assertion"]) -> int` 
3. `Config.get_assertion_verbosity() -> int`

bluetech commented:
> Can you clarify what you mean here, do you mean adding:

Yes, any of these would be fine by me.

nicoddemus commented:
One nice effect is that it would cut down on the number of things exposed publicly, specially 2) in my suggestion above: we can by default return the default on "unknown" verbosity types, to make the life of plugin authors easier: using a `VerbosityType` enum has many advantages, but one disadvantage is that plugin authors that need to support multiple pytest versions need to do some juggling to see if they have access to `VerbosityType.X`.

Suppose in the future we add `VerbosityType.Rubles` in pytest 9.0, plugins would have to juggle a bit to support pytest 7.x+ up to 9:

```python
if hasattr(config, "output_verbosity") and hasattr(pytest, "VerbosityType") and hasattr(pytest.VerbosityType, "Rubles"):
    rubles_verbosity = config.output_verbosity.get(pytest.VerbosityType, Rubles)
else:
    rubles_verbosity = config.option.verbose
```

While using option 2) above users can go with:

```python
if hasattr(config, "get_verbosity"):
    rubles_verbosity = config.get_verbosity("rubles")
else:
    rubles_verbosity = config.option.verbose
```

plannigan commented:
The forward comparability argument for pivoting back towards strings makes sense to me. I should have time this weekend to implement that and the absorption of `OutputVerbosity` into `Config`.

nicoddemus commented:
Thanks @plannigan, appreciate the patience and willing to change the implementation. 👍

nicoddemus commented on changelog/11387.feature.rst:
```diff
@@ -0,0 +1,5 @@
+
+See :ref:`Fine-grained verbosity <pytest.fine_grained_verbosity>` for more details.
+
+For plugin authors, :attr:`config.output_verbosity <pytest.Config.get_verbosity>` can be used to retrieve the verbosity level for a specific verbosity type.
```
```suggestion
For plugin authors, :attr:`config.get_verbosity <pytest.Config.get_verbosity>` can be used to retrieve the verbosity level for a specific verbosity type.
```

nicoddemus commented on testing/test_assertion.py:
```diff
@@ -13,27 +13,68 @@
+        config = mock_config(verbose=TestMockConfig.SOME_VERBOSITY_LEVEL)
+
+        with pytest.raises(KeyError):
+            config.get_verbosity("--- NOT A VERBOSITY LEVEL ---")
```
We do test all the pieces individually, which is great, but I think we should also include an integration test (using `pytester`) in this file, that runs pytest with different ` verbosity_assertions` levels and checks that the output differs accordingly.

plannigan replied:
Thanks for reminding me to do this. It made me notice that I had missed `saferepr`'s max size.

plannigan commented:
Looks like one of the jobs failed with a file system error in an area I didn't touch. I don't have permissions to retry the job.

nicoddemus commented:
> Looks like one of the jobs failed with a file system error in an area I didn't touch. I don't have permissions to retry the job.

Started it again. 👍

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,66 @@ def getvalueorskip(self, name: str, path=None):
+
+        return int(level)
+
+    @staticmethod
```
The name `_ini_name` is now too general, we should change it to `_verbosity_ini_name` or (my preference) inline it.

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,66 @@ def getvalueorskip(self, name: str, path=None):
+    _KNOWN_VERBOSITY_TYPES: Final = {VERBOSITY_ASSERTIONS}
+    _VERBOSITY_INI_DEFAULT = "auto"
+
+    def get_verbosity(self, verbosity_type: Optional[str] = None) -> int:
```
The docstring should explain what `None`/the default means.

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,66 @@ def getvalueorskip(self, name: str, path=None):
+            verbosity_type is None
+            or verbosity_type not in Config._KNOWN_VERBOSITY_TYPES
+        ):
+            return global_level
```
Let's add a unit test which covers this branch.

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,66 @@ def getvalueorskip(self, name: str, path=None):
+        return f"verbosity_{verbosity_type}"
+
+    @staticmethod
+    def _add_ini(parser: "Parser", verbosity_type: str, help: str) -> None:
```
Same here, the name `_add_ini` is too general now.

Also, IMO this function doesn't add much value, I think it would be better to reduce the indirection and inline it. But since it's private then it's OK if you want to keep it.

plannigan replied:
I prefer to keep it because it will avoid duplication as more types are added. It also ensures that the same default is used across verbosity types. If someone adds an Nth type, but uses "global", then `get_verbosity()` will raise an exception.

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,66 @@ def getvalueorskip(self, name: str, path=None):
+
+        level = self.getini(Config._ini_name(verbosity_type))
+
+        if level == Config._VERBOSITY_INI_DEFAULT:
```
The `auto` value should be documented.

Also, why did you choose "auto"?

plannigan replied:
Added some documentation. Also, it was the value used in [this related PR](https://github.com/pytest-dev/pytest/pull/10588/files#diff-0e2d32c88d3fa7edb466f8344dae556d52632e608e7414371f0537ce9dcc3a88R31) on the same topic. I'm open to changing it if there is a better suggestion.

bluetech commented on testing/test_assertion.py:
```diff
@@ -1836,3 +1877,54 @@ def test_comparisons_handle_colors(
+    pytester.makeini(
+        f"""
+        [pytest]
+        {_Config._ini_name(_Config.VERBOSITY_ASSERTIONS)} = 2
```
Write the real name here; as an integration test it should tell us if the actual thing the user writes regresses.

bluetech commented on testing/test_assertion.py:
```diff
@@ -1836,3 +1877,54 @@ def test_comparisons_handle_colors(
+        {_Config._ini_name(_Config.VERBOSITY_ASSERTIONS)} = 2
+        """
+    )
+    result = pytester.runpytest(str(p))
```
```suggestion
    result = pytester.runpytest(p)
```

plannigan replied:
Addressed, but in my defense, I copied this from other examples in the file :smile:.

bluetech reviewed (changes requested):
Haven't had time for a full review again, but left some comments, please take a look.

bluetech commented on doc/en/reference/reference.rst:
```diff
@@ -1000,7 +1000,6 @@ Stash
     :show-inheritance:
     :members:
 
-
```
Better to keep it

bluetech commented on src/_pytest/assertion/truncate.py:
```diff
@@ -1,15 +1,15 @@
+from _pytest.config import Config
 from _pytest.nodes import Item
 
-
```
Better to keep it

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,79 @@ def getvalueorskip(self, name: str, path=None):
+        assert isinstance(global_level, int)
+        if (
+            verbosity_type is None
+            or verbosity_type not in Config._KNOWN_VERBOSITY_TYPES
```
I think it would be better to avoid this check, so that plugins can add their own verbosity types. We would instead check if the ini exists.

plannigan replied:
I like the idea of checking the ini data instead (one less thing to manually manage). 

In a [previous conversation](https://github.com/pytest-dev/pytest/pull/11473#discussion_r1369828592), there was a conversation about plugins creating their own levels. I think I'm still in the "pytest can enumerate a few buckets" camp. @nicoddemus Do you have any thoughts? If we did allow plugins to create types, `_add_verbosity_ini` would have to be made public and some more documentation updates.

nicoddemus replied:
I think we can keep it simple for now, making this internal only; if the need comes later, we can discuss how to make this public -- better to err on the side of caution, as it is easy to make something public later, rather than making something public and then regretting it later.

bluetech commented on src/_pytest/config/__init__.py:
```diff
@@ -1650,6 +1651,79 @@ def getvalueorskip(self, name: str, path=None):
+    #: Verbosity type for failed assertions (see :confval:`verbosity_assertions`).
+    VERBOSITY_ASSERTIONS: Final = "assertions"
+    _KNOWN_VERBOSITY_TYPES: Final = {VERBOSITY_ASSERTIONS}
+    _VERBOSITY_INI_DEFAULT = "auto"
```
```suggestion
    _VERBOSITY_INI_DEFAULT: Final = "auto"
```

bluetech reviewed (changes requested):
Thanks for the update @plannigan, please see my comment on `_KNOWN_VERBOSITY_TYPES`. Other than that it looks good!

bluetech reviewed (approved):
LGTM! I know this feature will make a lot of people happy :)

I'll let @nicoddemus do the final review/merge.

nicoddemus commented:
Thanks everyone!
